fix set-service failing test

SteveL-MSFT opened the pull request:
missed -ErrorAction Stop
changed new-service to return error when given unsupported startuptype

Fix https://github.com/PowerShell/PowerShell/issues/4800
Fix https://github.com/PowerShell/PowerShell/issues/4803
<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

SteveL-MSFT committed:
[feature]

missed -ErrorAction Stop

iSazonov commented:
Blocked by #4803.

SteveL-MSFT committed:
[feature]

made cmdlet write error when using unsupported startuptype
re-enabled test

markekraus commented:
@SteveL-MSFT something should probably be done about [Set-Service.Tests.ps1#L25](https://github.com/PowerShell/PowerShell/blob/fb60c47a1ecbca2a6f287e0626a372497112f0a3/test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1#L25). That will continue to fail on Linux and macOS as the literal will still be parsed even though the `IT` is being skipped, and `Get-Service` is unavailable on Linux/macOS.

SteveL-MSFT commented:
@markekraus I see.  will fix.

SteveL-MSFT committed:
[feature]
fix test so that a script block only runs if It is run

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2015,10 +2015,11 @@ protected override void BeginProcessing()
-                        break;
+                        WriteNonTerminatingError(StartupType.ToString(), DisplayName, Name,
+                            new ArgumentException(), "CouldNotNewService",
+                            ServiceResources.UnsupportedStartupType,
```
There is another instance of `Assert(((ServiceStartMode)(-1)) == StartupType, "bad StartupType"` in `Set-Service`. It would be the best if all can be fixed here.

SteveL-MSFT replied:
will fix

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Management/resources/ServiceResources.resx:
```diff
@@ -201,4 +201,7 @@
     <value>The command cannot be used to configure the service '{0}' because access to computer '{1}' is denied. Run PowerShell as admin and run your command again.</value>
   </data>
+  <data name="UnsupportedStartupType" xml:space="preserve">
+    <value>The startup type '{0}' is not supported by New-Service.</value>
```
`by New-Service` needs to be changed since the same applies to `set-service` as well.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -163,8 +166,7 @@ Describe "Set/New-Service cmdlet tests" -Tags "Feature", "RequireAdminOnWindows"
         },
-        # This test case fails due to #4803. Disabled for now.
-        # @{name = 'badstarttype'; parameter = "StartupType"; value = "System"},
+        @{name = 'badstarttype'; parameter = "StartupType"; value = "System"},
```
Should we add "Boot" too? 
Should we check valid and invalid values from `System.ServiceProcess.ServiceStartMode`? If yes it is better refactor the tests. And for `Set-Service` too.

SteveL-MSFT replied:
will change

SteveL-MSFT committed:
[feature]
turned redundant code into helper function used by New/Set-Service
added tests for Set-Service and unsupported StartupTypes

SteveL-MSFT commented:
@iSazonov @daxian-dbw feedback addressed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1684,22 +1682,14 @@ protected override void ProcessRecord()
                             || (ServiceStartMode)(-1) != StartupType)
                         {
                             DWORD dwStartType = NativeMethods.SERVICE_NO_CHANGE;
```
I can make that change for New-Service, but for Set-Service, I need to use dwStartType on line 1697 so it has to be initialized.

iSazonov replied:
Thanks for clafiry. I think we should leave as is.
Closed.

daxian-dbw replied:
In `New-Service`, dwStartType is also used in line 2045 when calling `NativeMethods.CreateServiceW`.
It turns out the following code works:
```
static void Blah ()
{
    if (!int.TryParse("1", out int result))
    {
        Console.WriteLine("Failed");
    }

    Console.WriteLine(result);
}
```

daxian-dbw replied:
In `New-Service`, dwStartType is also used in line 2045 when calling `NativeMethods.CreateServiceW`.
It turns out the following code works:
```
static void Blah ()
{
    if (!int.TryParse("1", out int result))
    {
        Console.WriteLine("Failed");
    }

    Console.WriteLine(result);
}
```
It would be the best if we can do it in a consistent way in those 2 places.

SteveL-MSFT replied:
Addressed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2411,6 +2390,40 @@ internal static extern
+        /// <returns>
+        /// If a supported StartupType is provided, funciton returns true, otherwise false.
+        /// </returns>
+        internal static bool GetNativeStartupType(ServiceStartMode StartupType, out DWORD dwStartType)
```
Will change

SteveL-MSFT committed:
[feature]
address PR feedback

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1684,22 +1682,14 @@ protected override void ProcessRecord()
                         {
                             DWORD dwStartType = NativeMethods.SERVICE_NO_CHANGE;
-                            switch (StartupType)
+                            if ((ServiceStartMode)(-1) != StartupType &&
```
It seems `(ServiceStartMode)(-1) != StartupType &&` should be removed.

SteveL-MSFT replied:
Will change, but will need to change some other code to make this work.  Seems like `-1` shouldn't be needed at all.

SteveL-MSFT replied:
Actually, still need `-1` which will be treated as equivalent to `SERVICE_NO_CHANGE`

daxian-dbw replied:
By looking at the original code, for `New-Service`, `SERVICE_AUTO_START` would be used in case `StartupType` is -1. So it seems we cannot just use `dwStartType = NativeMethods.SERVICE_NO_CHANGE` for `(ServiceStartMode)(-1)`

SteveL-MSFT replied:
New-Service [defaults](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs#L1933) to Automatic whereas Set-Service [defaults](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs#L1511) to -1 to indicate no change.

daxian-dbw replied:
Please correct me if I'm wrong. The following snippet is the original code in `New-Service`. It indicates `dwStartType` should be `SERVICE_AUTO_START` when `StartupType == ((ServiceStartMode)(-1))`
```
DWORD dwStartType = NativeMethods.SERVICE_AUTO_START;
switch (StartupType)
{
    case ServiceStartMode.Automatic:
        dwStartType = NativeMethods.SERVICE_AUTO_START;
        break;
    case ServiceStartMode.Manual:
        dwStartType = NativeMethods.SERVICE_DEMAND_START;
        break;
    case ServiceStartMode.Disabled:
        dwStartType = NativeMethods.SERVICE_DISABLED;
        break;
    default:
        Diagnostics.Assert(
            ((ServiceStartMode)(-1)) == StartupType,
            "bad StartupType");
        break;
}
```
It's replaced with `if (!NativeMethods.TryGetNativeStartupType(StartupType, out DWORD dwStartType)) { // write error }` after the change, and `TryGetNativeStartupType` will set `dwStartType` to `SERVICE_NO_CHANGE` when `StartupType == ((ServiceStartMode)(-1))`.

So it looks to me changes the value of `dwStartType` in case of `StartupType == ((ServiceStartMode)(-1))`.

SteveL-MSFT replied:
I'll fix it so if it's (-1), it should default to Auto for New-Service

SteveL-MSFT committed:
[feature]
made StartupType(-1) equivalent to Win32 SERVICE_NO_CHANGE and moved to TryGetNativeStartupType

SteveL-MSFT commented:
@daxian-dbw feedback addressed

daxian-dbw reviewed (approved):
LGTM except for a non-blocking comment.

SteveL-MSFT commented:
@daxian-dbw can you review latest change?

daxian-dbw commented:
@SteveL-MSFT we cannot pass the value (-1) to `-StartupType` parameter. The parameter binding will fail because powershell cannot convert -1 to `[System.ServiceProcess.ServiceStartMode]`. That means `StartupType` in `New-Service` can never be `-1`, so we should be good without the last commit.

SteveL-MSFT commented:
@daxian-dbw yeah, I went through the code and for `New-Service`, StartupType defaults to Auto and can't ever be set to (-1), so I'll revert the last commit.

SteveL-MSFT commented:
Last commit was reverted.  We should be good to go.

iSazonov commented:
@SteveL-MSFT @daxian-dbw  Is the PR ready to merge?

daxian-dbw commented:
@iSazonov Yes, it's good to go.
